File: labs/daemon/modules/firewall/hotspot_repeater.go
package firewall

import (
	"fmt"
)

type HotspotRepeaterConfig struct {
	DownstreamInterface string `json:"downstream_interface"`
	UpstreamVpnInterface string `json:"upstream_vpn_interface"`
	DownstreamSubnet     string `json:"downstream_subnet"`
	ClampMssBytes       uint16 `json:"clamp_mss_bytes"`
}

func DefaultHotspotRepeaterConfig() HotspotRepeaterConfig {
	return HotspotRepeaterConfig{
		DownstreamInterface:  "wlan1",
		UpstreamVpnInterface: "tun0",
		DownstreamSubnet:     "192.168.43.0/24",
		ClampMssBytes:        1360,
	}
}

type HotspotNatRepeater struct {
	config HotspotRepeaterConfig
}

func NewHotspotNatRepeater(cfg HotspotRepeaterConfig) *HotspotNatRepeater {
	return &HotspotNatRepeater{config: cfg}
}

func (r *HotspotNatRepeater) GenerateIptablesCommands() []string {
	return []string{
		"echo 1 > /proc/sys/net/ipv4/ip_forward",
		fmt.Sprintf("iptables -A FORWARD -i %s -o %s -j ACCEPT", r.config.DownstreamInterface, r.config.UpstreamVpnInterface),
		fmt.Sprintf("iptables -A FORWARD -i %s -o %s -m state --state RELATED,ESTABLISHED -j ACCEPT", r.config.UpstreamVpnInterface, r.config.DownstreamInterface),
		fmt.Sprintf("iptables -t nat -A POSTROUTING -s %s -o %s -j MASQUERADE", r.config.DownstreamSubnet, r.config.UpstreamVpnInterface),
		fmt.Sprintf("iptables -t mangle -A FORWARD -p tcp --tcp-flags SYN,RST SYN -j TCPMSS --set-mss %d", r.config.ClampMssBytes),
	}
}

File: labs/control-ui/api-shadow/hotspot_repeater.ts
export interface HotspotConfig {
  downstreamIface?: string;
  upstreamVpnIface?: string;
  downstreamSubnet?: string;
  clampMssBytes?: number;
}

export class HotspotNatRepeater {
  public readonly downstreamIface: string;
  public readonly upstreamVpnIface: string;
  public readonly downstreamSubnet: string;
  public readonly clampMssBytes: number;

  constructor(cfg: HotspotConfig = {}) {
    this.downstreamIface = cfg.downstreamIface ?? 'wlan1';
    this.upstreamVpnIface = cfg.upstreamVpnIface ?? 'tun0';
    this.downstreamSubnet = cfg.downstreamSubnet ?? '192.168.43.0/24';
    this.clampMssBytes = cfg.clampMssBytes ?? 1360;
  }

  public generateIptables(): string[] {
    return [
      'echo 1 > /proc/sys/net/ipv4/ip_forward',
      `iptables -A FORWARD -i ${this.downstreamIface} -o ${this.upstreamVpnIface} -j ACCEPT`,
      `iptables -A FORWARD -i ${this.upstreamVpnIface} -o ${this.downstreamIface} -m state --state RELATED,ESTABLISHED -j ACCEPT`,
      `iptables -t nat -A POSTROUTING -s ${this.downstreamSubnet} -o ${this.upstreamVpnIface} -j MASQUERADE`,
      `iptables -t mangle -A FORWARD -p tcp --tcp-flags SYN,RST SYN -j TCPMSS --set-mss ${this.clampMssBytes}`
    ];
  }
}
